### DISCUSSION TITLE
Unique users folder

### DISCUSSION THREAD

**Discussion by volkankaban** [spatie/laravel-medialibrary#2912] [General]
I want to be sure, before purchasing Media Library, can I use with unique users folders like .uploads/{model_id}/ for each user? I want to give them only so they can keep all photos in one place. It's a quick solution for me because I can share a folder with them, and they can access it anytime?

unique users folder;
uploads/11/ *.*
uploads/12/ *.*
uploads/13/ *.*
uploads/14/ *.*
uploads/15/ *.*

users in DB;
users.id = media.model_id
11         = 11
12         = 12
13         = 13
14         = 14
15         = 15

CustomPathGenerator.php
<?php
namespace App\MediaLibrary;

use App\Models\Product;
use App\Models\Setting;
use App\Models\User;
use Spatie\MediaLibrary\MediaCollections\Models\Media;
use Spatie\MediaLibrary\Support\PathGenerator\PathGenerator;

class CustomPathGenerator implements PathGenerator
{
    public function getPath(Media $media): string
    {
        $path = '{PARENT_DIR}'.'/'.$media->model_id.'/';
		
        switch ($media->collection_name) {
            case User::PROFILE;
                return str_replace('{PARENT_DIR}', User::PROFILE, $path);
            case Product::Image;
                return str_replace('{PARENT_DIR}', Product::Image, $path);
            case Setting::PATH;
                return str_replace('{PARENT_DIR}', Setting::PATH, $path);
            case 'default';
                return '';
        }
    }

    /**
     * @param  Media  $media
     * @return string
     */
    public function getPathForConversions(Media $media): string
    {
        return $this->getPath($media).'thumbnails/';
    }

    public function getPathForResponsiveImages(Media $media): string
    {
        return $this->getPath($media).'rs-images/';
    }
}

**Comment by freekmurze** [comment]
You have also mailed support, and I've replied there.

**Reply by volkankaban** [reply]
Thank you very much for the response! The custom path generator method worked well! ->> /users/<user-id>/media/<media-id> 

I'm trying to set up a global function for all image uploads, and is there any way to change by default for all media filenames to 32 characters  md5 or uniqid() or str methods without changing media extensions .ext(strtolower)? 


    public function store($input)
    {
        try {
            DB::beginTransaction();
            $input['password'] = \Illuminate\Support\Facades\Hash::make($input['password']);
            $user = User::create($input);
            if (isset($input['profile']) && ! empty($input['profile'])) {
		$hashImg = md5(Hash::make($user));
		$pathext  = pathinfo($input['profile'], PATHINFO_EXTENSION);
                $user->addMedia($input['profile'])->usingName($hashImg)->usingFileName("$hashImg.$pathext")->toMediaCollection(User::PROFILE, config('app.storage'));
            }
            DB::commit();
            return true;
        } catch (\Exception $e) {
            DB::rollBack();
            throw new UnprocessableEntityHttpException($e->getMessage());
        }
    }

Result for filename: /users/22/media/1/bb488c88663529a3faa78468d3ee87ce.tmp  <<-- missing extension.

    
References:
https://github.com/spatie/laravel-medialibrary/blob/256eb5f086d80cfb9e9525401268ed164f8ebab7/docs/handling-uploads-with-media-library-pro/processing-uploads-on-the-server.md 
  
https://github.com/spatie/laravel-medialibrary/blob/c0d9bcdfdd954523f817295d2fde73401e1a8e91/tests/Support/PathGenerator/CustomPathGenerator.php 

Pro version introduction:
https://github.com/spatie/laravel-medialibrary/blob/256eb5f086d80cfb9e9525401268ed164f8ebab7/docs/handling-uploads-with-media-library-pro/introduction.md

**Reply by volkankaban** [reply]
I'm not sure about second response email, I got just one email; 

> > 
> Hi,
> ​
> ​Seems like you are on the right track, you indeed need to create a custom path generator like explained in our docs. You can test your path generator out with the free version of medialib.
> ​
> Just be aware that when delete media, medialib will delete the parent directory so create a directory like this
> ​
> /users/<user-id>/media/<media-id>
> ​
> ​and not this
> ​
> ​/users/<user-id>
> ​
> ​Kr,
> ​
> ​
> ​
> ​Freek
> >
